File: LeetCode/Maximum Length of Repeated Subarray/Maximum Length of Repeated Subarray(二分查找法).java
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

interface AlgoStrategy
{
	int execute();
}

class BinarySearchAlgo implements AlgoStrategy
{
	public BinarySearchAlgo(int[] a, int[] b)
	{
		this.a = a;
		this.b = b;
	}
	
	public int execute()
	{
		int low = 0;
		int high = Math.min(a.length, b.length);
		
		while (low < high)
		{
			int mid = (low + high + 1) >> 1;
			if (check(mid))
			{
				low = mid;
			}
			else 
			{
				high = mid - 1;
			}
		}
		
		return low;
	}
	
	private boolean check(int len)
	{
		Set<String> seenSet = new HashSet<>();
		
		for (int i = 0; i + len <= a.length; ++i)
		{
			seenSet.add(Arrays.toString(Arrays.copyOfRange(a, i, i + len)));
		}
		
		for (int i = 0; i + len <= b.length; ++i)
		{
			if (seenSet.contains(Arrays.toString(Arrays.copyOfRange(b, i, i + len))))
			{
				return true;
			}
		}
		
		return false;
	}
	
	private int[] a;
	private int[] b;
}

interface AlgoFactory<T>
{
	T create();
}

class AlgoFactoryImpl implements AlgoFactory<BinarySearchAlgo>
{
	public AlgoFactoryImpl(int[] a, int[] b)
	{
		this.a = a;
		this.b = b;
	}
	
	public BinarySearchAlgo create()
	{
		return new BinarySearchAlgo(a, b);
	}
	
	private int[] a;
	private int[] b;
}

public class Solution 
{
    public int findLength(int[] A, int[] B) 
    {
        AlgoFactory algoFactory = new AlgoFactoryImpl(A, B);
        AlgoStrategy algoStrategy = (AlgoStrategy)algoFactory.create();
        return algoStrategy.execute();
    }
}

File: LeetCode/Maximum Length of Repeated Subarray/Maximum Length of Repeated Subarray(二分查找法).go
package main

//时间复杂度O((m+n)*log(min(m,n))*min(m,n))
import (
	"fmt"
	"math"
	"strings"
)

func findLength(A []int, B []int) int {
	var algo AlgoStrategy
	algo = &BinarySearchAlgo{A, B}
	return algo.Execute()
}

type AlgoStrategy interface {
	Execute() int
}

type BinarySearchAlgo struct {
	a []int
	b []int
}

func (algo *BinarySearchAlgo) check(length int) bool {
	a := algo.a
	b := algo.b

	set := make(map[string]bool)
	for i := 0; i + length <= len(a); i++ {
		s := strings.Replace(strings.Trim(fmt.Sprint(a[i:i + length]), "[]"), " ", "", -1)
		set[s] = true
	}

	for i := 0; i + length <= len(b); i++ {
		s := strings.Replace(strings.Trim(fmt.Sprint(b[i:i + length]), "[]"), " ", "", -1)
		if _, ok := set[s]; ok {
			return true
		}
	}

	return false
}

func (algo *BinarySearchAlgo) Execute() int {
	a := algo.a
	b := algo.b

	high := int(math.Min(float64(len(a)), float64(len(b))))
	low := 0

	for low < high {
		mid := (low + high + 1) >> 1
		if algo.check(mid) {
			low = mid
		} else {
			high = mid - 1
		}
	}

	return low
}
